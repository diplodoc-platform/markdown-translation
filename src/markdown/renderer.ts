import MarkdownIt from 'markdown-it';
import {
    mdRenderer,
    MarkdownRendererParams,
    MarkdownRendererEnv,
} from '@diplodoc/markdown-it-markdown-renderer';

// configure with diplodoc plugins
// @ts-ignore
import meta from 'markdown-it-meta';
import notes from '@doc-tools/transform/lib/plugins/notes';

import markdownHandlers, {MarkdownHandlersState} from './handlers';
import hooks, {HooksParameters} from './hooks';

export type MarkdownRendererState = MarkdownHandlersState;

export type RenderParameters = {
    skeleton: string;
    translations: Map<string, string>;
} & DiplodocParameters;

export type DiplodocParameters = {
    lang?: string;
};

function render(parameters: RenderParameters) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const md = new MarkdownIt('commonmark', {html: true}) as HooksParameters['markdownit'];

    const {handlers, initState} = markdownHandlers.generate(parameters);
    const markdownHooks = hooks.generate({markdownit: md});
    const mdOptions: MarkdownRendererParams<MarkdownHandlersState> = {
        handlers,
        initState,
        hooks: markdownHooks.hooks,
    };
    const diplodocOptions = {
        lang: parameters.lang ?? 'ru',
    };

    const env: MarkdownRendererEnv = {
        source: parameters.skeleton.split('\n'),
    };

    md.use(mdRenderer, mdOptions);

    // diplodoc plugins
    md.use(meta);
    md.use(notes, diplodocOptions);

    return md.render(parameters.skeleton, env);
}

function validParameters(parameters: RenderParameters) {
    const {skeleton, translations} = parameters;

    const skeletonCondition = skeleton !== undefined;
    const translationsCondition = translations !== undefined;

    const conditions = [
        skeletonCondition,
        translationsCondition,
        validDiplodocParameters(parameters),
    ];

    return conditions.reduce((a, v) => a && v, true);
}

function validDiplodocParameters(parameters: DiplodocParameters) {
    const {lang} = parameters;

    return lang === undefined || lang === 'ru' || lang === 'en';
}

export {render, validParameters, validDiplodocParameters};
export default {render, validParameters, validDiplodocParameters};
