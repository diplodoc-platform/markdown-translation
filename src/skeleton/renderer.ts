import MarkdownIt from 'markdown-it';
import {
    mdRenderer,
    MarkdownRendererParams,
    MarkdownRendererEnv,
} from '@diplodoc/markdown-it-markdown-renderer';

// configure with diplodoc plugins
// @ts-ignore
import meta from 'markdown-it-meta';
// @ts-ignore
import sup from 'markdown-it-sup';
import notes from '@doc-tools/transform/lib/plugins/notes';
import cut from '@doc-tools/transform/lib/plugins/cut';
import checkbox from '@doc-tools/transform/lib/plugins/checkbox';

import skeletonHandlers, {SkeletonHandlersState} from './handlers';
import hooks, {HooksParameters} from './hooks';

export type SkeletonRendererState = SkeletonHandlersState;

export type RenderParameters = {
    markdown: string;
} & DiplodocParameters;

export type DiplodocParameters = {
    lang?: string;
};

function render(parameters: RenderParameters) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const {markdown, lang} = parameters;

    const md = new MarkdownIt({html: true}) as HooksParameters['markdownit'];
    const env: MarkdownRendererEnv = {source: markdown.split('\n')};

    const {handlers, initState} = skeletonHandlers.generate();

    const skeletonHooks = hooks.generate({markdownit: md});
    const mdOptions: MarkdownRendererParams<SkeletonHandlersState> = {
        handlers,
        initState,
        hooks: skeletonHooks.hooks,
    };
    const diplodocOptions = {
        lang: lang ?? 'ru',
    };

    md.use(mdRenderer, mdOptions);

    // diplodoc plugins
    md.use(meta, diplodocOptions);
    md.use(notes, diplodocOptions);
    md.use(cut, diplodocOptions);
    md.use(sup, diplodocOptions);
    md.use(checkbox, diplodocOptions);

    return md.render(markdown, env);
}

function validParameters(parameters: RenderParameters) {
    const {markdown, lang} = parameters;

    const markdownCondition = markdown !== undefined;
    const langCondition = lang === undefined || lang === 'ru' || lang === 'en';
    const conditions = [markdownCondition, langCondition];

    return conditions.reduce((a, v) => a && v, true);
}

export {render, validParameters};
export default {render, validParameters};
