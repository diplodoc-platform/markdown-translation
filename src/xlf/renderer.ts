import MarkdownIt from 'markdown-it';
import {customRenderer, CustomRendererParams} from '@diplodoc/markdown-it-custom-renderer';

// configure with diplodoc plugins
// @ts-ignore
import meta from 'markdown-it-meta';
import notes from '@doc-tools/transform/lib/plugins/notes';

import {template} from './generator';
import rules, {XLFRulesState} from './rules';
import hooks, {HooksParameters} from './hooks';

export type XLFRendererState = XLFRulesState;

export type RenderParameters = {
    markdown: string;
} & template.TemplateParameters &
    DiplodocParameters;

export type DiplodocParameters = {
    lang?: string;
};

function render(parameters: RenderParameters) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const wrapper = template.generate(parameters);

    const xlfRenderer = new MarkdownIt('commonmark', {html: true}) as HooksParameters['markdownit'];
    const xlfRules = rules.generate(wrapper);
    const xlfHooks = hooks.generate({template: wrapper.template, markdownit: xlfRenderer});
    const xlfOptions: CustomRendererParams<XLFRendererState> = {
        rules: xlfRules.rules,
        hooks: xlfHooks.hooks,
        initState: xlfRules.initState,
    };
    const diplodocOptions = {
        lang: parameters.lang ?? 'ru',
    };

    xlfRenderer.use(customRenderer, xlfOptions);

    // diplodoc plugins
    xlfRenderer.use(meta);
    xlfRenderer.use(notes, diplodocOptions);

    return xlfRenderer.render(parameters.markdown);
}

function validParameters(parameters: RenderParameters) {
    const {lang, markdown} = parameters;

    const markdownCondition = markdown !== undefined;
    const langCondition = lang === undefined || lang === 'ru' || lang === 'en';
    const conditions = [template.validParameters(parameters), markdownCondition, langCondition];

    return conditions.reduce((a, v) => a && v, true);
}

export {render, validParameters};
export default {render, validParameters};
