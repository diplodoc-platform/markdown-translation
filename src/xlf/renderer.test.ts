import {render, RenderParameters} from './renderer';
import {
    markdown,
    markdownWithVariables,
    markdownWithConditionals,
    markdownWithLoops,
    markdownWithFilters,
    markdownWithFunctions,
    markdownWithMeta,
    markdownWithNotes,
    markdownWithCuts,
} from 'src/__fixtures__';

describe('smoke', () => {
    test('it works', () => {
        const parameters = {
            markdown: '',
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
            skeletonPath: 'file.skl.md',
            markdownPath: 'file.md',
        };

        render(parameters);
    });
});

describe('validates parameters', () => {
    it('works with valid parameters', () => {
        const parameters = {
            markdown: '',
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
            skeletonPath: 'file.skl.md',
            markdownPath: 'file.md',
        };

        render(parameters);
    });

    it('throws on invalid parameters', () => {
        const parameters = {
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
            skeletonPath: 'file.skl.md',
            markdownPath: 'file.md',
        } as RenderParameters;

        const invalidLang = {
            markdown: '',
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
            skeletonPath: 'file.skl.md',
            markdownPath: 'file.md',
            lang: 'xx',
        } as RenderParameters;

        expect(() => render(parameters)).toThrow();
        expect(() => render(invalidLang)).toThrow();
    });
});

describe('xlf rendering', () => {
    it('renders empty xlf document for empty markdown', () => {
        const parameters = {
            markdown: '',
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
            skeletonPath: 'file.skl.md',
            markdownPath: 'file.md',
        };

        const rendered = render(parameters);
        expect(rendered).toMatchSnapshot();
    });

    it('renders trans-units from text tokens', () => {
        const parameters = {
            markdown,
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
            markdownPath: 'text.md',
            skeletonPath: 'text.skl.md',
        };

        const rendered = render(parameters);
        expect(rendered).toMatchSnapshot();
    });

    // naive approach leaves variables inside of text as is
    it('handles markdown with liquid variables', () => {
        const parameters = {
            markdown: markdownWithVariables,
            source: {
                language: 'en',
                locale: 'US' as const,
            },
            target: {
                language: 'ru',
                locale: 'RU' as const,
            },
            markdownPath: 'text.md',
            skeletonPath: 'text.skl.md',
        };

        const rendered = render(parameters);
        expect(rendered).toMatchSnapshot();
    });

    it('handles markdown with liquid filters', () => {
        const parameters = {
            markdown: markdownWithFilters,
            source: {
                language: 'en',
                locale: 'US' as const,
            },
            target: {
                language: 'ru',
                locale: 'RU' as const,
            },
            markdownPath: 'text.md',
            skeletonPath: 'text.skl.md',
        };

        const rendered = render(parameters);
        expect(rendered).toMatchSnapshot();
    });

    it('handles markdown with liquid functions', () => {
        const parameters = {
            markdown: markdownWithFunctions,
            source: {
                language: 'en',
                locale: 'US' as const,
            },
            target: {
                language: 'ru',
                locale: 'RU' as const,
            },
            markdownPath: 'text.md',
            skeletonPath: 'text.skl.md',
        };

        const rendered = render(parameters);
        expect(rendered).toMatchSnapshot();
    });

    it('handles markdown with liquid conditionals', () => {
        const parameters = {
            markdown: markdownWithConditionals,
            source: {
                language: 'en',
                locale: 'US' as const,
            },
            target: {
                language: 'ru',
                locale: 'RU' as const,
            },
            markdownPath: 'text.md',
            skeletonPath: 'text.skl.md',
        };

        const rendered = render(parameters);
        expect(rendered).toMatchSnapshot();
    });

    it('handles markdown with liquid loops', () => {
        const parameters = {
            markdown: markdownWithLoops,
            source: {
                language: 'en',
                locale: 'US' as const,
            },
            target: {
                language: 'ru',
                locale: 'RU' as const,
            },
            markdownPath: 'text.md',
            skeletonPath: 'text.skl.md',
        };

        const rendered = render(parameters);
        expect(rendered).toMatchSnapshot();
    });

    it('handles markdown with meta', () => {
        const parameters = {
            markdown: markdownWithMeta,
            source: {
                language: 'en',
                locale: 'US' as const,
            },
            target: {
                language: 'ru',
                locale: 'RU' as const,
            },
            markdownPath: 'text.md',
            skeletonPath: 'text.skl.md',
        };

        const rendered = render(parameters);
        expect(rendered).toMatchSnapshot();
    });

    it('handles markdown with notes', () => {
        const parameters = {
            markdown: markdownWithNotes,
            source: {
                language: 'en',
                locale: 'US' as const,
            },
            target: {
                language: 'ru',
                locale: 'RU' as const,
            },
            markdownPath: 'text.md',
            skeletonPath: 'text.skl.md',
        };

        const rendered = render(parameters);
        expect(rendered).toMatchSnapshot();
    });

    it('handles markdown with cuts', () => {
        const parameters = {
            markdown: markdownWithCuts,
            source: {
                language: 'en',
                locale: 'US' as const,
            },
            target: {
                language: 'ru',
                locale: 'RU' as const,
            },
            markdownPath: 'text.md',
            skeletonPath: 'text.skl.md',
        };

        const rendered = render(parameters);
        expect(rendered).toMatchSnapshot();
    });
});
