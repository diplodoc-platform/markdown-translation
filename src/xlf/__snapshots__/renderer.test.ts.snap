// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xlf rendering renders empty xlf document for empty markdown 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<xliff xmlns=\\"urn:oasis:names:tc:xliff:document:1.2\\" version=\\"1.2\\">
  <file original=\\"file.md\\" source-language=\\"ru-RU\\" target-language=\\"en-US\\" datatype=\\"markdown\\">
    <header>
      <skeleton>
        <external-file href=\\"file.skl.md\\"></external-file>
      </skeleton>
    </header>
    <body>
    </body>
  </file>
</xliff>
"
`;

exports[`xlf rendering renders trans-units from text tokens 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<xliff xmlns=\\"urn:oasis:names:tc:xliff:document:1.2\\" version=\\"1.2\\">
  <file original=\\"text.md\\" source-language=\\"ru-RU\\" target-language=\\"en-US\\" datatype=\\"markdown\\">
    <header>
      <skeleton>
        <external-file href=\\"text.skl.md\\"></external-file>
      </skeleton>
    </header>
    <body>
      <trans-unit id=\\"1\\">
        <source>Заголовок 1</source>
      </trans-unit>
      <trans-unit id=\\"2\\">
        <source>Параграф</source>
      </trans-unit>
      <trans-unit id=\\"3\\">
        <source>Первый элемент списка</source>
      </trans-unit>
      <trans-unit id=\\"4\\">
        <source>Второй элемент списка</source>
      </trans-unit>
      <trans-unit id=\\"5\\">
        <source>Заголовок 2</source>
      </trans-unit>
      <trans-unit id=\\"6\\">
        <source>Параграф содержащий</source>
      </trans-unit>
      <trans-unit id=\\"7\\">
        <source>жирный текст</source>
      </trans-unit>
      <trans-unit id=\\"8\\">
        <source>и</source>
      </trans-unit>
      <trans-unit id=\\"9\\">
        <source>курсив</source>
      </trans-unit>
      <trans-unit id=\\"10\\">
        <source>А ещё ключевые особенности структуры проекта освещают чрезвычайно интересные особенности картины в целом.</source>
      </trans-unit>
      <trans-unit id=\\"11\\">
        <source>Предложения складываются в абзацы — и вы наслаждетесь очередным бредошедевром.</source>
      </trans-unit>
      <trans-unit id=\\"12\\">
        <source>текст внутри цитаты</source>
      </trans-unit>
      <trans-unit id=\\"13\\">
        <source>Первый элемент нумерованного списка</source>
      </trans-unit>
      <trans-unit id=\\"14\\">
        <source>Второй элемент нумерованного списка</source>
      </trans-unit>
      <trans-unit id=\\"15\\">
        <source>Текст ссылки.</source>
      </trans-unit>
      <trans-unit id=\\"16\\">
        <source>Всё ещё текст ссылки.</source>
      </trans-unit>
      <trans-unit id=\\"17\\">
        <source>Описание ссылки.</source>
      </trans-unit>
      <trans-unit id=\\"18\\">
        <source>Всё ещё описание ссылки.</source>
      </trans-unit>
      <trans-unit id=\\"19\\">
        <source>Альтернативный текст картинки.</source>
      </trans-unit>
      <trans-unit id=\\"20\\">
        <source>Всё ещё альт. текст картинки.</source>
      </trans-unit>
      <trans-unit id=\\"21\\">
        <source>Титул картинки.</source>
      </trans-unit>
      <trans-unit id=\\"22\\">
        <source>Всё ещё титул картинки.</source>
      </trans-unit>
    </body>
  </file>
</xliff>
"
`;
